### ISSUE TITLE
how to run in optimized mode ?

### ISSUE THREAD

**Issue by zapashcanon** [ocaml-wasm/wasm_of_ocaml#104]
Hi,

I'm making some benchmarks and I'd like to be sure to run wsoo with the best possible performances. For now I'm doing:

```shell-session
$ ocamlc ./kb.ml
$ wasm_of_ocaml compile --opt=3 a.out
$ node a.js
```

But the results seem slower than what I would expect on some tests:

```shell-session
$ ./bench.sh
*** Running Knuth-Bendix
Wasocaml (node):            real 1.00 user 1.35 sys 0.06
Wasocaml + wasm-opt (node): real 0.99 user 1.37 sys 0.04
OCaml native:               real 0.67 user 0.66 sys 0.00
OCaml bytecode:             real 2.77 user 2.76 sys 0.00
js_of_ocaml (node):         real 24.62 user 25.39 sys 0.16
wasm_of_ocaml (node):       real 8.18 user 8.50 sys 0.06
```

Is there anything else I should do to get better performances from wsoo? From my understanding it calls Binaryen itself and I should not have to do it myself, is this right?

Thanks!

**Comment by OlivierNicole** [comment]
> Is there anything else I should do to get better performances from wsoo? From my understanding it calls Binaryen itself and I should not have to do it myself, is this right?

Yes. You can pass `--debug binaryen` to know exactly how Binaryen is invoked.

Interesting result! Four times slower than bytecode is certainly not typical. The first thing I would do would be to use Chrome’s profiler to inspect both the Wasocaml-compiled and the wasm_of_ocaml-compiled program and try to determine where the slowdown is located with wasm_of_ocaml. If you want to do it, or want to share the source program, we would be grateful!

**Comment by vouillon** [comment]
This is an exception-heavy benchmark. The implementation of exceptions in the V8 engine is quite slow.
I guess Wasocaml does not use Wasm exceptions but implements its own exception mechanism.
